articles/libraries/auth0-android/save-and-refresh-tokens.md

```diff
@@ -0,0 +1,117 @@
+
+```java
+Auth0 auth0 = new Auth0(this);
+AuthenticationAPIClient apiClient = new AuthenticationAPIClient(auth0);
+CredentialsManager manager = new CredentialsManager(apiClient, new SharedPreferencesStorage(this));
+```
+
+### Current State of the Authentication
+By asking for credentials existence upon launch and clearing them on logout, the manager can be used to manage the state of the authentication inside your app.
+
+Stored credentials are considered valid if they have not expired or can be refreshed. Check if a user has already logged in:
+
+```java
+boolean loggedIn = manager.hasValidCredentials();
+```
+
+When you want to log out the user from your app, remove the stored credentials and direct them to the login screen.
+
+```java
+manager.clearCredentials();
+//Show login screen
+```
+
+### Obtain Credentials
+Because the credentials may need to be refreshed against Auth0 Servers, this method is a-synchronic. Pass a callback implementation where you'd like to receive the credentials. Credentials returned by this method upon success are always valid.
+
+```java
+manager.getCredentials(new BaseCallback<Credentials, CredentialsManagerException>() {
+    @Override
+    public void onSuccess(Credentials credentials) {
+        //Use credentials
+    }
+
+    @Override
+    public void onFailure(CredentialsManagerException error) {
+        //No credentials were previously saved or they couldn't be refreshed
+    }
+});
+```
+
+::: note
+If the credentials were renewed, the manager will automatically save the new value for you.
+:::
+
+### Save new Credentials
+You can save the credentials obtained upon log in in the manager. If they contain a valid `refresh_token` they will be refreshed automatically if they expire, in other case, they will be deemed invalid.
+
+```java
+manager.saveCredentials(credentials);
+```
+
+
+## Using the SecureCredentialsManager class
+
+### Setup
+Setup is a little different to the CredentialsManager class. Create a new instance by passing a valid android `Context`, an `AuthenticationAPIClient` and a `Storage` implementation.
+
+```java
+Auth0 auth0 = new Auth0(this);
+AuthenticationAPIClient apiClient = new AuthenticationAPIClient(auth0);
+SecureCredentialsManager manager = new SecureCredentialsManager(this, apiClient, new SharedPreferencesStorage(this));
+```
+
+The methods to obtain, save, check for existence and clearing the credentials are the ones explained before.
+
+
+### Pre-Authenticate the User
+This class can require the user to authenticate before being able to obtain the credentials by using the device's configured Lock Screen. If the Lock Screen Security is set to something different than PIN, Pattern, Password or Fingerprint, this feature won't be available. You need to call the method below to enable the authentication. Pass a valid `Activity` context, a request code, and 2 optional Strings to use as title and description for the Lock Screen.
```

Comment:
This class provides optional functionality for additional authentication using the device's configured Lock Screen.